engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeVisualisationsTest.scala

```diff
@@ -1665,4 +1665,192 @@ class RuntimeVisualisationsTest
       context.executionComplete(contextId)
     )
   }
+
+  it should "run visualisation expression catching error" in {
+    val contextId       = UUID.randomUUID()
+    val requestId       = UUID.randomUUID()
+    val visualisationId = UUID.randomUUID()
+    val moduleName      = "Enso_Test.Test.Main"
+    val metadata        = new Metadata
+
+    val idMain = metadata.addItem(46, 14)
+
+    val code =
+      """from Standard.Builtins import all
+        |
+        |main =
+        |    Error.throw 42
+        |""".stripMargin.linesIterator.mkString("\n")
+    val contents = metadata.appendToCode(code)
+    val mainFile = context.writeMain(contents)
+
+    // create context
+    context.send(Api.Request(requestId, Api.CreateContextRequest(contextId)))
+    context.receive shouldEqual Some(
+      Api.Response(requestId, Api.CreateContextResponse(contextId))
+    )
+
+    // Open the new file
+    context.send(
+      Api.Request(Api.OpenFileNotification(mainFile, contents))
+    )
+    context.receiveNone shouldEqual None
+
+    // push main
+    val item1 = Api.StackItem.ExplicitCall(
+      Api.MethodPointer(moduleName, "Enso_Test.Test.Main", "main"),
+      None,
+      Vector()
+    )
+    context.send(
+      Api.Request(requestId, Api.PushContextRequest(contextId, item1))
+    )
+    context.receive(3) should contain theSameElementsAs Seq(
+      Api.Response(requestId, Api.PushContextResponse(contextId)),
+      TestMessages.error(
+        contextId,
+        idMain,
+        Api.ExpressionUpdate.Payload.DataflowError(Seq())
+      ),
+      context.executionComplete(contextId)
+    )
+
+    // attach visualisation
+    context.send(
+      Api.Request(
+        requestId,
+        Api.AttachVisualisation(
+          visualisationId,
+          idMain,
+          Api.VisualisationConfiguration(
+            contextId,
+            moduleName,
+            "x -> x.catch_primitive _.to_text"
+          )
+        )
+      )
+    )
+    val attachVisualisationResponses = context.receive(3)
+    attachVisualisationResponses should contain allOf (
+      Api.Response(requestId, Api.VisualisationAttached()),
+      context.executionComplete(contextId)
+    )
+    val Some(data) = attachVisualisationResponses.collectFirst {
+      case Api.Response(
+            None,
+            Api.VisualisationUpdate(
+              Api.VisualisationContext(
+                `visualisationId`,
+                `contextId`,
+                `idMain`
+              ),
+              data
+            )
+          ) =>
+        data
+    }
+    data.sameElements("42".getBytes) shouldBe true
+  }
+
+  it should "run visualisation expression propagating panic" in {
```

Comment:
Personally, it's not super clear to me why we actually want to ensure this behavior (through writing a test for it); is there some reason why we *need* that a visualization attached to a panicking node is ever evaluated at all? wouldn't it be perfectly fine also if the visualization was *not* executed at all in such a case (and thus would not propagate the panic IIUC)?

With that said, from my side I'm marking this as accepted, you most certainly know much more about this than me!

Reply:
Thanks for the review :heart: 

The preprocessor behavior on error expressions was not tested, thus it was not clear for me how it will perform in this situation. For example, I had a wrong assumption that since a panic expression does not evaluate to a value, IDE cannot call a visualization expression on it (no value - nothing to run the expression on). These test cases clarified and documented the behavior.

The case of not running the visualization at all can be an optimization, but I'm not sure if we want to do this right now. For now, I'd like to keep things simple and the visualization processor behavior is the same for every node.